prime-router/src/main/kotlin/fhirengine/engine/FHIRDestinationFilter.kt

```diff
@@ -0,0 +1,561 @@
+     * return [defaultResponse]. If the filter doesn't pass the results are logged on the [report] for
+     * that specific [filterType].
+     * @param filters Filters that will be evaluated
+     * @param bundle FHIR Bundle that will be evaluated
+     * @param report Report object passed for logging purposes
+     * @param receiver Receiver of the report passed for logging purposes
+     * @param filterType Type of filter passed for logging purposes
+     * @param defaultResponse Response returned if the filter is null or empty
+     * @param reverseFilter Optional flag used to reverse the result of the filter evaluation
+     * @param focusResource Starting point for the evaluation, can be [bundle] if checking from root
+     * @param useOr Optional flag used to allow conditions to be evaluated with "or" instead of "and"
+     * @return Boolean indicating if the bundle passes the filter or not
+     *        Result will be negated if [reverseFilter] is true
+     */
+    internal fun evaluateFilterAndLogResult(
+        filters: ReportStreamFilter,
+        bundle: Bundle,
+        reportId: ReportId,
+        actionHistory: ActionHistory,
+        receiver: Receiver,
+        filterType: ReportStreamFilterType,
+        defaultResponse: Boolean,
+        reverseFilter: Boolean = false,
+        focusResource: Base = bundle,
+        useOr: Boolean = false,
+    ): Boolean {
+        val evaluationFunction = if (useOr) ::evaluateFilterConditionAsOr else ::evaluateFilterConditionAsAnd
+        val (passes, failingFilterName) = evaluationFunction(
+            filters,
+            bundle,
+            defaultResponse,
+            reverseFilter,
+            focusResource
+        )
+        if (!passes) {
+            logFilterResults(
+                failingFilterName ?: "unknown",
+                bundle,
+                reportId,
+                actionHistory,
+                receiver,
+                filterType,
+                focusResource
+            )
+        }
+        return passes
+    }
+
+    /**
+     * Takes a [bundle] and [filter] and optionally a [focusResource], evaluates if the bundle passes all
+     * the filter conditions, or when [reverseFilter] is true, evaluates if the bundle doesn't pass at least one of the
+     * filter conditions.
+     * @param filter Filter that will be evaluated
+     * @param bundle FHIR Bundle that will be evaluated
+     * @param defaultResponse result when there are no filter conditions in the list
+     * @param reverseFilter Optional flag used to reverse the result of the filter evaluation
+     * @param focusResource Starting point for the evaluation, can be [bundle] if checking from root
+     * @return Pair: Boolean indicating if the bundle passes the filter or not
+     *         and String to use when logging the filter result
+     */
+    internal fun evaluateFilterConditionAsAnd(
```

Comment:
Might make sense to simplify this function a lot.  

- The default response can be dropped since it's always false
- Do we really need the concept of a reverse filter here?
- focusResource will always be bundle right?